### PR TITLE
Conditional stemming for 'persian' analyzer

### LINKED ISSUE #113050: Lucene 10 "persian" analyzer now stems by default
Lucene 10s PersianAnalyzer that we expose via the "persian" now [adds a final stemming filter](https://github.com/apache/lucene/blob/main/lucene/MIGRATE.md#persianstemfilter-is-added-to-persiananalyzer-lucene-10312) in https://github.com/apache/lucene/pull/540. It looks like this is the default now and can only be turned off by providing an exclude-list or creating a custom analyzer without the stemming filter.

We temporarily changed the language filter documentation that show how the analyzers are internally build [on the lucene_snapshot_10 branch](https://github.com/elastic/elasticsearch/commit/455980df9f949073a3d0de7be10d7cc1c53c055a) but need to discuss and decide if we want to regard this as a breaking change or somehow restore the 9x behavior in ES.

### PR DESCRIPTION
**Pull request by cbuescher**
The 'persian' analyzer for Lucene 10 comes with PersianStemFilter as the last token filter by default. In order to maintain compatibility for old indices, we only use the new analyzer for newly created indices but configure a legacy analyzer with the old behaviour for older index versions.

Closes #113050

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/analysis/Analysis.java [hunk 1]
```diff
@@ -83,6 +84,8 @@ public class Analysis {
     private static final DeprecationLogger DEPRECATION_LOGGER = DeprecationLogger.getLogger(Analysis.class);
     private static final Logger logger = LogManager.getLogger(Analysis.class);
 
+    public static final NodeFeature PERSIAN_ANALYZER_WITH_STEMMER = new NodeFeature("persian_analyzer_with_stemmer");
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/analysis/Analysis.java:87]: I originally wanted to put this into the PersianAnalyzerProvider but I cannot see analysis-common from RestFeatures where I need to reference it. Still not sure if its overkill to add a "capability" for this but saw it also done around modifications in Highlighting (see RestFeatures) so I think its okay. Let me know if you know better options.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/analysis/Analysis.java:87]: The purpose of these is to distinguish between node deployments. Since we know this will be deployed with Lucene 10 changes, let's just have a "Lucene 10 upgrade" feature and use it. 

We can also use that lucene 10 feature for any other bwc things that need changing.

**Review by benwtrent — APPROVED** [review]
I like the testing and most of the changes. I do think the node features we need to add is a "Lucene 10 Release" feature instead of individual ones that are required for the lucene 10 release.

File: server/src/main/java/org/elasticsearch/index/analysis/Analysis.java (hunk 1 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/analysis/Analysis.java]: Moved to `SearchFeatures`

File: server/src/main/java/org/elasticsearch/search/SearchFeatures.java [hunk 2]
```diff
@@ -16,8 +16,11 @@
 import java.util.Set;
 
 public final class SearchFeatures implements FeatureSpecification {
+
+    public static final NodeFeature LUCENE_10_0_0_UPGRADE = new NodeFeature("lucene_10_upgrade");
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/SearchFeatures.java:20]: Turns out @benwtrent had a very good intuition. The day after we chatted about it, i had to introduce this to address the profile yaml test failures. See https://github.com/elastic/elasticsearch/commit/c0646368f782decdd9d1109c25e12b36746c6e1a . Heads up this node feature is already available in the branch.

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/SearchFeatures.java]: Removed from the diff now.

**Comment by cbuescher** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by cbuescher** [comment]
[CI Failures on 8.16-bwc](https://gradle-enterprise.elastic.co/s/bioa3xhxcx56e/tests/overview?outcome=FAILED) are [issues coming from "main"](https://github.com/elastic/elasticsearch/issues?q=is%3Aissue%20state%3Aopen%2040_routing_partition_size) so I'm inclined to merge this and wait for a fix/awaitsFix with the next upstream merge

**Comment by cbuescher** [comment]
@benwtrent I added a [changelog entry](https://github.com/elastic/elasticsearch/pull/113482/commits/11003e40e7276d5d544ed8cfcb7c59e9989f6b92), let me know if that reads alright to you.

File: server/src/main/java/org/elasticsearch/index/analysis/AnalyzerProvider.java [hunk 3]
```diff
@@ -12,12 +12,12 @@
 import org.apache.lucene.analysis.Analyzer;
 import org.elasticsearch.injection.guice.Provider;
 
-public interface AnalyzerProvider<T extends Analyzer> extends Provider<T> {
+public interface AnalyzerProvider<T extends Analyzer> extends Provider<Analyzer> {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/analysis/AnalyzerProvider.java:15]: Is this change required?

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalyzerProvider.java:15]: Yes, because PersianAnalyzerProvider now either creates a PersianAnalyzer or an alternative StopwordAnalyzerBase implementation based on the index version. Since PersianAnalyzer (like almost all other analyzers in Lucene) is final, we cannot change any of its behavior. We don't need the interface type anywhere anyway, so we can simply user Analyzer here.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalyzerProvider.java:15]: I still don't get it, sorry. It looks odd to me that we need to keep the generic type yet specify `Provider<Analyzer>`, perhaps there's something I am missing.

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalyzerProvider.java:15]: That option didn't occur to me, probably because `StopwordAnalyzerBase` is abstract but you're right, that's an option I will try. 
And yes, I would have liked to remove the generic Type from AnalyzerProvider entirely because it seems unused but that would indeed be a larger change (though possibly mostly mechanical), so I avoided that.

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalyzerProvider.java]: See [57b0686](https://github.com/elastic/elasticsearch/pull/113482/commits/57b068610229c445f729489534d8cdf23d49561b) for an update

**Comment by leemthompo** [comment]
@cbuescher is this PR relevant to the serverless changelog? [FYI this question is based on [9.0 breaking changes](https://staging-website.elastic.co/docs/release-notes/elasticsearch/breaking-changes)]

**Comment by cbuescher** [comment]
> is this PR relevant to the serverless changelog?

Yes I think so.
